processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -300,10 +317,172 @@ public int compare(Object lhs, Object rhs)
     );
   }
 
-  public Ordering<Row> getRowOrdering(final boolean granular)
+  private boolean validateAndGetForceLimitPushDown()
+  {
+    final boolean forcePushDown = getContextBoolean(GroupByQueryConfig.CTX_KEY_FORCE_LIMIT_PUSH_DOWN, false);
+    if (forcePushDown) {
+      if (!(limitSpec instanceof DefaultLimitSpec)) {
+        throw new IAE("When forcing limit push down, a limit spec must be provided.");
+      }
+
+      if (((DefaultLimitSpec) limitSpec).getLimit() == Integer.MAX_VALUE) {
+        throw new IAE("When forcing limit push down, the provided limit spec must have a limit.");
+      }
+
+      for (OrderByColumnSpec orderBySpec : ((DefaultLimitSpec) limitSpec).getColumns()) {
+        if (OrderByColumnSpec.getPostAggIndexForOrderBy(orderBySpec, postAggregatorSpecs) > -1) {
+          throw new UnsupportedOperationException("Limit push down when sorting by a post aggregator is not supported.");
+        }
+      }
+    }
+    return forcePushDown;
+  }
+
+  public boolean willApplyLimitPushDown()
+  {
+    final boolean forceLimitPushDown = validateAndGetForceLimitPushDown();
+
+    if (limitSpec instanceof DefaultLimitSpec) {
+      DefaultLimitSpec defaultLimitSpec = (DefaultLimitSpec) limitSpec;
+
+      // If only applying an orderby without a limit, don't try to push down
+      if (defaultLimitSpec.getLimit() == Integer.MAX_VALUE) {
+        return false;
+      }
+
+      if (forceLimitPushDown) {
+        return true;
+      }
+
+      // If the sorting order only uses columns in the grouping key, we can always push the limit down
+      // to the buffer grouper without affecting result accuracy
+      boolean sortHasNonGroupingFields = DefaultLimitSpec.sortingOrderHasNonGroupingFields(
+          limitSpec,
+          getDimensions()
+      );
+
+      return !sortHasNonGroupingFields;
+    }
+
+    return false;
+  }
+
+  private Ordering<Row> getRowOrderingForPushDown(
```

Comment:
When limit push down is applied, the partial results would be sorted by the ordering specified by the limit/order spec (unlike non-push down case where the results always use the default natural ascending order), so when merging these partial result streams, the merge needs to use the same ordering to get correct results